README.md
# AnyBar: OS X menubar status indicator

AnyBar is a small indicator for your menubar that does one simple thing: it displays a colored dot. What the dot means and when to change it is up to you.

<img src="screenshot.png?raw=true" />

## Download

Version 0.2.3:

<a href="https://github.com/tonsky/AnyBar/releases/download/0.2.3/AnyBar-0.2.3.zip"><img src="AnyBar/Images.xcassets/AppIcon.appiconset/icon_128x128@2x.png?raw=true" style="width: 128px;" width=128/></a>

Or using [Homebrew Cask](https://github.com/Homebrew/homebrew-cask):

    brew install --cask anybar

## Support us

<a href="https://patreon.com/tonsky" target="_blank"><img src="./anybar_patreon.png"></a>

## Usage

AnyBar is controlled via a UDP port (1738 by default). Before any commands can be sent, AnyBar.app must be launched:

```sh
open -a AnyBar
```

Once launched, you may send it a message to change the style of the dot:

```sh
echo -n "black" | nc -4u -w0 localhost 1738
```

The following default commands change the style of the dot:

| Command       | Preview                                                             |
|---------------|---------------------------------------------------------------------|
| `white`       | <img src="AnyBar/Resources/white@2x.png?raw=true" width=15 />       |
| `red`         | <img src="AnyBar/Resources/red@2x.png?raw=true" width=15 />         |
| `orange`      | <img src="AnyBar/Resources/orange@2x.png?raw=true" width=15 />      |
| `yellow`      | <img src="AnyBar/Resources/yellow@2x.png?raw=true" width=15 />      |
| `green`       | <img src="AnyBar/Resources/green@2x.png?raw=true" width=15 />       |
| `cyan`        | <img src="AnyBar/Resources/cyan@2x.png?raw=true" width=15 />        |
| `blue`        | <img src="AnyBar/Resources/blue@2x.png?raw=true" width=15 />        |
| `purple`      | <img src="AnyBar/Resources/purple@2x.png?raw=true" width=15 />      |
| `black`       | <img src="AnyBar/Resources/black@2x.png?raw=true" width=15 />       |
| `question`    | <img src="AnyBar/Resources/question@2x.png?raw=true" width=15 />    |
| `exclamation` | <img src="AnyBar/Resources/exclamation@2x.png?raw=true" width=15 /> |
| `filled`      | <img src="AnyBar/Resources/filled@2x.png?raw=true" width=15 />      |
| `hollow`      | <img src="AnyBar/Resources/hollow@2x.png?raw=true" width=15 />      |

`black` and `white` always has black or white fill. On Big Sur, where text color of menubar might change depending on the wallpaper, you might want to use `filled` and `hollow` instead. They are inverted when menubar changes its appearance.

To quit, send `quit`.

## Alternative clients

Bash alias:

```sh
$ function anybar { echo -n $1 | nc -4u -w0 localhost ${2:-1738}; }

$ anybar red
$ anybar green 1739
```

Zsh with completion:

- [wookayin/anybar-zsh](https://github.com/wookayin/anybar-zsh)

Fish shell with completion:

- [matchai/anybar-fish](https://github.com/matchai/anybar-fish)

Go:

- [justincampbell/anybar](https://github.com/justincampbell/anybar)
- [johntdyer/anybar-go](https://github.com/johntdyer/anybar-go)

Node:

- [rumpl/nanybar](https://github.com/rumpl/nanybar)
- [sindresorhus/anybar](https://github.com/sindresorhus/anybar)
- [snippet by skibz](https://github.com/tonsky/AnyBar/issues/11)

PHP:

- [2bj/Phanybar](https://github.com/2bj/Phanybar)

Java:

- [cs475x/AnyBar4j](https://github.com/cs475x/AnyBar4j)

Python:

- [philipbl/pyanybar](https://github.com/philipbl/pyAnyBar)

Ruby:

- [davydovanton/AnyBar_rb](https://github.com/davydovanton/AnyBar_rb)

Rust:

- [urschrei/rust_anybar](https://github.com/urschrei/rust_anybar)
- [Feliix42/anybar-rs](https://github.com/Feliix42/anybar-rs)

Nim:

- [rgv151/anybar.nim](https://github.com/rgv151/anybar.nim)

Erlang:

- [kureikain/ebar](https://github.com/kureikain/ebar)

C:

- [onderweg/anybar-cli](https://github.com/onderweg/anybar-cli)

C#:

- [jenyayel/anybar-client](https://github.com/jenyayel/anybar-client)

Crystal:
- [davydovanton/AnyBar_cr](https://github.com/davydovanton/AnyBar_cr)

Emacs: 

- [rmuslimov/anybar.el](https://gist.github.com/rmuslimov/2d74cacd5e0ae827663e)
- [tie-rack/anybar-el](https://github.com/tie-rack/anybar-el) (Also on [Melpa](http://melpa.org/#/anybar))

AppleScript:

```
tell application "AnyBar" to set image name to "blue"

tell application "AnyBar" to set current to get image name as Unicode text
display notification current
```

Alfred:

- [https://github.com/raguay/MyAlfred](https://github.com/raguay/MyAlfred/blob/master/Alfred%204/AnyBar%20Workflow.alfredworkflow)

## Integrations

- Webpack build status plugin [roman01la/anybar-webpack](https://github.com/roman01la/anybar-webpack)
- boot-clj task [tonsky/boot-anybar](https://github.com/tonsky/boot-anybar)
- Idea plugin [denofevil/AnyBarIdea](https://github.com/denofevil/AnyBarIdea)
- Anybar-based CLI journal [Andrew565/anybar-icon-journal](https://github.com/Andrew565/anybar-icon-journal)
- Command monitoring [rvirani1/with_anybar](https://github.com/rvirani1/with_anybar)
- Monitor commands automatically, across several iterm tabs [stacycurl/anybar_bash](https://github.com/stacycurl/anybar-bash)
- Extension for ipython/jupyter/ipython notebook [ermakovpetr/ipython-anybar](https://github.com/ermakovpetr/ipython-anybar)

## Running multiple instances

You can run several instances of AnyBar as long as they listen on different ports. Use the `ANYBAR_PORT` environment variable to change the port and `open -na` to run several instances:

```sh
ANYBAR_PORT=1738 open -na AnyBar
ANYBAR_PORT=1739 open -na AnyBar
ANYBAR_PORT=1740 open -na AnyBar
```

## Environment variables to specify a title and the initial color of the dot

A title can be set to distinguish dots in the menubar:

```sh
ANYBAR_PORT=1738 ANYBAR_TITLE=First  open -na AnyBar
ANYBAR_PORT=1739 ANYBAR_TITLE=Second open -na AnyBar
ANYBAR_PORT=1740 ANYBAR_TITLE=Third  open -na AnyBar
```

And the initial color of the dot can also be set:

```sh
ANYBAR_INIT=blue open -na AnyBar
```

## Custom images

AnyBar can detect and use local custom images stored in the `~/.AnyBar` directory. For example, if you have a `~/.AnyBar/square@2x.png` image, send `square` to port 1738 and it will be displayed. Images should be 19×19 pixels for standard resolution, and 38x38 pixels for retina (@2x).

## Ports

- Ubuntu Unity [limpbrains/somebar](https://github.com/limpbrains/somebar)
- i3wm with i3pystatus [enkore/i3pystatus](https://github.com/enkore/i3pystatus)
- Windows 10 [PavelStefanov/NoteBar](https://github.com/PavelStefanov/NoteBar)
- Emacs [plexus/.../emybar.el](https://github.com/plexus/plexmacs/blob/master/emybar/emybar.el)

## License

Copyright © 2015 Nikita Prokopov

Licensed under Eclipse Public License (see [LICENSE](LICENSE)).


CHANGELOG.md
### 0.2.3 / Feb 18, 2021

- Added `filled` and `hollow`, removed alternative versions of `black` and `white` #84

### 0.2.2 / Nov 3, 2020

- Specify starting color via `ANYBAR_INIT` #80 #81 thx @Gira-X

### 0.2.1 / Aug 28, 2020

- Correct version metadata

### 0.2.0 / Aug 26, 2020

- Specify app tooltip via `ANYBAR_TITLE` #59 #64 thx @mynameismiek @andrewsjg

### 0.1.4

- Bigger dots
- Render “black” on dark menubar as empty circle and “white” on dark as filled circle (#55)
- Compiled for OS X 10.11

### 0.1.3

- AppleScript support (PR #8, thx [Oleg Kertanov](https://github.com/okertanov))

### 0.1.2

- Dark mode support. In dark mode AnyBar will first check for `<image>_alt@2x.png` or `<image>_alt.png` image first, then falls back to `<image>.png`
- Support for Mavericks actually works

### 0.1.1

- Support for Mavericks (PR #2, thx [Oleg Kertanov](https://github.com/okertanov))
- Support for custom images via ~/.AnyBar (PR #1, thx [Paul Boschmann](https://github.com/pboschmann))


AnyBar/AppDelegate.m
//
//  AppDelegate.m
//  AnyBar
//
//  Created by Nikita Prokopov on 14/02/15.
//  Copyright (c) 2015 Nikita Prokopov. All rights reserved.
//

#import "AppDelegate.h"

@interface AppDelegate()

@property (weak, nonatomic) IBOutlet NSWindow *window;
@property (strong, nonatomic) NSStatusItem *statusItem;
@property (strong, nonatomic) GCDAsyncUdpSocket *udpSocket;
@property (strong, nonatomic) NSString *imageName;
@property (assign, nonatomic) int udpPort;
@property (assign, nonatomic) NSString *appTitle;

@end

@implementation AppDelegate

-(void)applicationDidFinishLaunching:(NSNotification *)aNotification {
    _udpPort = -1;
    _imageName = [self readStringFromEnvironmentVariable:@"ANYBAR_INIT" usingDefault:@"hollow"];
    self.statusItem = [self initializeStatusBarItem];
    [self setImage:_imageName];

    @try {
        _udpPort = [self getUdpPort];
        _udpSocket = [self initializeUdpSocket: _udpPort];
        _appTitle = [self readStringFromEnvironmentVariable:@"ANYBAR_TITLE" usingDefault:nil];
        _statusItem.toolTip = _appTitle == nil ? [NSString stringWithFormat:@"AnyBar @ %d", _udpPort] : _appTitle;
    }
    @catch(NSException *ex) {
        NSLog(@"Error: %@: %@", ex.name, ex.reason);
        _statusItem.image = [NSImage imageNamed:@"exclamation@2x.png"];
        [_statusItem.image setTemplate:NO];
    }
    @finally {
        NSString *portTitle = [NSString stringWithFormat:@"UDP port: %@", _udpPort >= 0 ? [NSNumber numberWithInt:_udpPort] : @"unavailable"];
        NSMenu *menu = [[NSMenu alloc] init];
        
        if (_appTitle != nil)
            [menu addItemWithTitle:_appTitle action:nil keyEquivalent:@""];
        [menu addItemWithTitle:portTitle action:nil keyEquivalent:@""];
        [menu addItemWithTitle:@"Quit" action:@selector(terminate:) keyEquivalent:@""];

        _statusItem.menu = menu;
    }
}

-(void)applicationWillTerminate:(NSNotification *)aNotification {
    [self shutdownUdpSocket: _udpSocket];
    _udpSocket = nil;

    [[NSStatusBar systemStatusBar] removeStatusItem:_statusItem];
    _statusItem = nil;
}

-(int) getUdpPort {
    int port = [self readIntFromEnvironmentVariable:@"ANYBAR_PORT" usingDefault:@"1738"];

    if (port < 0 || port > 65535) {
        @throw([NSException exceptionWithName:@"Argument Exception"
                            reason:[NSString stringWithFormat:@"UDP Port range is invalid: %d", port]
                            userInfo:@{@"argument": [NSNumber numberWithInt:port]}]);

    }

    return port;
}

-(GCDAsyncUdpSocket*)initializeUdpSocket:(int)port {
    NSError *error = nil;
    GCDAsyncUdpSocket *udpSocket = [[GCDAsyncUdpSocket alloc]
                                    initWithDelegate:self
                                    delegateQueue:dispatch_get_main_queue()];

    [udpSocket bindToPort:port error:&error];
    if (error) {
        @throw([NSException exceptionWithName:@"UDP Exception"
                            reason:[NSString stringWithFormat:@"Binding to %d failed", port]
                            userInfo:@{@"error": error}]);
    }

    [udpSocket beginReceiving:&error];
    if (error) {
        @throw([NSException exceptionWithName:@"UDP Exception"
                            reason:[NSString stringWithFormat:@"Receiving from %d failed", port]
                            userInfo:@{@"error": error}]);
    }

    return udpSocket;
}

-(void)shutdownUdpSocket:(GCDAsyncUdpSocket*)sock {
    if (sock != nil) {
        [sock close];
    }
}

-(void)udpSocket:(GCDAsyncUdpSocket *)sock didReceiveData:(NSData *)data
      fromAddress:(NSData *)address withFilterContext:(id)filterContext {
    [self processUdpSocketMsg:sock withData:data fromAddress:address];
}

-(NSImage*)tryImage:(NSString *)path {
    NSFileManager *fileManager = [NSFileManager defaultManager];
    if ([fileManager fileExistsAtPath:path])
        return [[NSImage alloc] initWithContentsOfFile:path];
    else
        return nil;
}

-(NSString*)bundledImagePath:(NSString *)name {
    return [[NSBundle mainBundle] pathForResource:name ofType:@"png"];
}

-(NSString*)homedirImagePath:(NSString *)name {
    return [NSString stringWithFormat:@"%@/%@/%@.png", NSHomeDirectory(), @".AnyBar", name];
}

-(void)setImage:(NSString*) name {
    NSImage *image = nil;
    image = [self tryImage:[self homedirImagePath:[name stringByAppendingString:@"@2x"]]];
    if (!image)
        image = [self tryImage:[self homedirImagePath:name]];
    if (!image)
        image = [self tryImage:[self bundledImagePath:[name stringByAppendingString:@"@2x"]]];
    if (!image)
        image = [self tryImage:[self bundledImagePath:name]];
    if (!image) {
        NSLog(@"Cannot find image '%@'", name);
        image = [self tryImage:[self bundledImagePath:@"question@2x"]];
        _statusItem.image = image;
        [_statusItem.image setTemplate:NO];
    } else {
        _statusItem.image = image;
        if ([name isEqualToString:@"filled"] || [name isEqualToString:@"hollow"])
            [_statusItem.image setTemplate:YES];
        else
            [_statusItem.image setTemplate:NO];
        _imageName = name;
    }
}

-(void)processUdpSocketMsg:(GCDAsyncUdpSocket *)sock withData:(NSData *)data
    fromAddress:(NSData *)address {
    NSString *msg = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];

    if ([msg isEqualToString:@"quit"])
        [[NSApplication sharedApplication] terminate:nil];
    else
        [self setImage:msg];
}

-(NSStatusItem*) initializeStatusBarItem {
    NSStatusItem *statusItem = [[NSStatusBar systemStatusBar] statusItemWithLength:NSVariableStatusItemLength];
//    statusItem.image = [NSImage imageNamed:@"white@2x.png"];
    statusItem.highlightMode = YES;
    return statusItem;
}

-(int) readIntFromEnvironmentVariable:(NSString*) envVariable usingDefault:(NSString*) defStr {
    int intVal = -1;

    NSString *envStr = [[[NSProcessInfo processInfo]
                         environment] objectForKey:envVariable];
    if (!envStr) {
        envStr = defStr;
    }

    NSNumberFormatter *nFormatter = [[NSNumberFormatter alloc] init];
    nFormatter.numberStyle = NSNumberFormatterDecimalStyle;
    NSNumber *number = [nFormatter numberFromString:envStr];

    if (!number) {
        @throw([NSException exceptionWithName:@"Argument Exception"
                            reason:[NSString stringWithFormat:@"Parsing integer from %@ failed", envStr]
                            userInfo:@{@"argument": envStr}]);

    }

    intVal = [number intValue];

    return intVal;
}

-(NSString*) readStringFromEnvironmentVariable:(NSString*) envVariable usingDefault:(NSString*) defStr {
    NSString *envStr = [[[NSProcessInfo processInfo]
                         environment] objectForKey:envVariable];
    
    if (!envStr) {
        envStr = defStr;
    }
    
    return envStr;
}

-(id) osaImageBridge {
    NSLog(@"OSA Event: %@ - %@", NSStringFromSelector(_cmd), _imageName);

    return _imageName;
}


-(void) setOsaImageBridge:(id)imgName {
    NSLog(@"OSA Event: %@ - %@", NSStringFromSelector(_cmd), imgName);

    _imageName = (NSString *)imgName;

    [self setImage:_imageName];
}

@end

